### PR TITLE
feat: update and test soil id algorithm

### PR DESCRIPTION
**Pull request by shrouxm**
## Description
Updates the soil ID algorithm and adds unit tests for our resolver code.

### Checklist
- [x] New tests added

### Related Issues
Progress on https://github.com/techmatters/soil-id-algorithm/issues/8

### CODE REVIEW

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        )
+        prev_depth = bottom_depth
+
+    return SoilIdSoilData(
+        slope=soil_match["site"]["siteData"]["slope"], depth_dependent_data=depth_dependent_data
+    )
+
+
+def resolve_ecological_site(ecological_site: dict):
+    if ecological_site["ecoclassid"] == "" or ecological_site["ecoclassid"][0] == "":
```
Comment (paulschreiber) [inline terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:85]:
```suggestion
    if not ecological_site["ecoclassid"] or not ecological_site["ecoclassid"][0]:
```

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        distance_to_nearest_map_unit_m=site_data["minCompDistance"],
+        match=resolve_soil_match_info(soil_id["score_loc"], soil_id["rank_loc"]),
+        soil_info=resolve_soil_info(soil_match),
+    )
+
+
+def resolve_location_matches_from_soil_id_result(soil_list_json: dict):
+    matches = []
+    for match in soil_list_json["soilList"]:
+        if match["id"]["rank_loc"] != "Not Displayed":
```
Comment (paulschreiber) [inline terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:134]: question: how many special values like `Not Displayed` are there? Should they be in constants?

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    SoilInfo,
+    SoilMatchInfo,
+    SoilSeries,
+)
+from apps.soil_id.models.depth_dependent_soil_data import DepthDependentSoilData
+from apps.soil_id.models.soil_data import SoilData
+
+
+def resolve_texture(texture: str | float):
+    if not isinstance(texture, str) or texture == "":
```
Comment (paulschreiber) [inline terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:45]:
```suggestion
    if not isinstance(texture, str) or not texture:
```

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+
+def parse_surface_cracks(surface_cracks: SoilData.SurfaceCracks):
+    if surface_cracks is None:
+        return None
+    return surface_cracks == SoilData.SurfaceCracks.DEEP_VERTICAL_CRACKING
+
+
+def parse_rank_soils_input_data(data: SoilIdInputData):
+    # TODO: pass in values for elevation and bedrock
```
Comment (paulschreiber) [inline terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:207]: question: is this work tracked?

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    elif rock_fragment_volume == DepthDependentSoilData.RockFragmentVolume.VOLUME_0_1:
+        return "0-1%"
+    elif rock_fragment_volume == DepthDependentSoilData.RockFragmentVolume.VOLUME_1_15:
+        return "1-15%"
+    elif rock_fragment_volume == DepthDependentSoilData.RockFragmentVolume.VOLUME_15_35:
+        return "15-35%"
+    elif rock_fragment_volume == DepthDependentSoilData.RockFragmentVolume.VOLUME_35_60:
+        return "35-60%"
+    else:
+        return ">60%"
```
Comment (paulschreiber) [inline terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:191]: suggestion: use constants for the return values.

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py (hunk 4 above)
Comment (shrouxm) [reply terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:207]: no, i don't think so, but this isn't new code

File: Makefile [hunk 6]
```diff
@@ -2,7 +2,7 @@ DC_ENV ?= dev
 DC_FILE_ARG = -f docker-compose.$(DC_ENV).yml
 DC_RUN_CMD = docker compose $(DC_FILE_ARG) run --quiet-pull --rm web
 
-SCHEMA_BUILD_CMD = $(DC_RUN_CMD) python terraso_backend/manage.py graphql_schema --schema apps.graphql.schema.schema --out=-.graphql
+SCHEMA_BUILD_CMD = $(DC_RUN_CMD) python terraso_backend/manage.py graphql_schema --schema apps.graphql.schema.schema.schema --out=-.graphql
```
[the remark this answers is not in the document]
Comment (shrouxm) [reply Makefile:5]: previously:
- the first schema was a folder name, which implicitly imported the `__init__.py` under that folder, and the second schema was the name of an exported value in that file
now:
- the first schema is a folder name, the second schema is a file name, and the third is the exported value

i had to do this in order to fix an import cycle

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py (hunk 1 above)
Comment (shrouxm) [reply terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:85]: i disagree that not being explicit about what we're testing for is better here

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py (hunk 3 above)
Comment (shrouxm) [reply terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:45]: same as above, i like being explicit about what i am testing for

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py (hunk 2 above)
Comment (shrouxm) [reply terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:134]: maybe, but i already made a big push on code quality in this PR with the reorganization + unit testing that i wasn't sure if there was time for so i'm hesitant to keep trying to make this code really nice. i should file a tech debt issue to do this kind of cleanup though

File: terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py (hunk 5 above)
Comment (shrouxm) [reply terraso_backend/apps/soil_id/graphql/soil_id/resolvers.py:191]: see: https://github.com/techmatters/terraso-backend/pull/1329#discussion_r1646771217

**Review by tm-ruxandra — APPROVED** [review]

**Pull request merged**